dds/DCPS/XTypes/DynamicVwrite.cpp

```diff
@@ -0,0 +1,1042 @@
+      if (log_level >= LogLevel::Warning) {
+        ACE_ERROR((LM_WARNING, "(%P|%t) WARNING: vwrite_item: get_descriptor for sequence failed (%C)\n",
+                   retcode_to_string(rc)));
+      }
+      return;
+    }
+    DDS::DynamicType_var elem_type = get_base_type(seq_td->element_type());
+    const TypeKind elem_kind = elem_type->get_kind();
+    TypeKind treat_elem_as;
+    if (get_equivalent_kind(elem_type, treat_elem_as) != DDS::RETCODE_OK) {
+      return;
+    }
+    // Try writing the whole primitive sequence. If fails, fall back to write element one by one.
+    if (is_primitive(treat_elem_as) &&
+        vwrite_primitive_array_i(vw, value, id, treat_elem_as, elem_kind, true) == DDS::RETCODE_OK) {
+      break;
+    }
+  }
+  case TK_STRUCTURE:
+  case TK_UNION:
+  case TK_ARRAY: {
+    DDS::DynamicData_var member_data;
+    rc = value->get_complex_value(member_data, id);
+    if (!check_rc(rc, id, treat_as, "vwrite_item")) {
+      return;
+    }
+    begin_member_helper(vw, params, rc, containing_tk);
+    if (rc == DDS::RETCODE_OK) {
+      vwrite(vw, member_data);
+    } else {
+      vw.write_absent_value();
+    }
+    break;
+  }
+  default:
+    if (log_level >= LogLevel::Warning) {
+      ACE_ERROR((LM_WARNING, "(%P|%t) WARNING: vwrite_item: Unsupported type %C\n",
+                 typekind_to_string(treat_as)));
+    }
+    return;
+  }
+
+  if (containing_tk == TK_STRUCTURE) {
+    vw.end_struct_member();
+  } else if (containing_tk == TK_UNION) {
+    vw.end_union_member();
+  }
+}
+
+void vwrite_member(ValueWriter&vw, DDS::DynamicData_ptr value,
+                   const DDS::MemberDescriptor_var& md, DDS::TypeKind containing_tk)
+{
+  const DDS::MemberId id = md->id();
+  const DDS::DynamicType_var member_type = XTypes::get_base_type(md->type());
+  VWriterMemberParam params(id, md->is_must_understand() || md->is_key(), md->name(),
+                            md->is_optional(), true);
+  vwrite_item(vw, value, id, member_type, containing_tk, &params);
+}
+
+void vwrite_struct(ValueWriter& vw, DDS::DynamicData_ptr value, const DDS::DynamicType_var& dt)
+{
+  DDS::TypeDescriptor_var type_desc;
+  DDS::ReturnCode_t rc = dt->get_descriptor(type_desc);
+  if (rc != DDS::RETCODE_OK) {
+    if (log_level >= LogLevel::Warning) {
+      ACE_ERROR((LM_WARNING, "(%P|%t) WARNING: vwrite_struct: get_descriptor failed (%C)\n",
+                 retcode_to_string(rc)));
+    }
+    return;
+  }
+
+  const Extensibility extensibility = XTypes::dds_to_opendds_ext(type_desc->extensibility_kind());
+  vw.begin_struct(extensibility);
+  for (CORBA::ULong i = 0; i < dt->get_member_count(); ++i) {
+    DDS::DynamicTypeMember_var dtm;
+    rc = dt->get_member_by_index(dtm, i);
+    if (rc != DDS::RETCODE_OK) {
+      if (log_level >= LogLevel::Warning) {
+        ACE_ERROR((LM_WARNING, "(%P|%t) WARNING: vwrite_struct: get_member_by_index %u failed (%C)\n",
+                   i, retcode_to_string(rc)));
+      }
+      return;
```

Comment:
As the first step, I think the value writer interface should change to return something like `bool` for every method now.
When an error like this happens, the `Xcdr2ValueWriter` object has an invalid state because it expects the matching `end_struct`, and so we can't reuse it.

Reply:
Returning a bool makes sense.